src/rebalancers/libraries/RebalanceLogic.sol

```diff
@@ -0,0 +1,270 @@
+     *       To maximise the liquidity of the new position, the balances after the swap should approximate it as close as possible to not have any leftovers.
+     *    b) The "Current Ratio" (R_current) is the ratio of the current token balances, it is calculated as follows:
+     *       R_current = balance1 / [balance0 * sqrtPrice² + balance1].
+     *
+     * 3) From R_target and R_current we can finally derive the direction of the swap, amountIn and amountOut.
+     *    If R_current is smaller than R_target, we have to swap an amount of token0 to token1, and vice versa.
+     *    amountIn and amountOut can be found by solving the following equalities:
+     *      a) The ratio of the token balances after the swap equal the "Target Ratio".
+     *      b) The swap between token0 and token1 is done in the pool itself,
+     *         taking into account fees, but ignoring slippage (-> sqrtPrice remains constant).
+     *
+     *    If R_current < R_target (swap token0 to token1):
+     *      a) R_target = [amount1 + amoutOut] / [(amount0 - amountIn) * sqrtPrice² + (amount1 + amoutOut)].
+     *      b) amountOut = (1 - fee) * amountIn * sqrtPrice².
+     *         => amountOut = [(R_target - R_current) * (amount0 * sqrtPrice² + amount1)] / [1 + R_target * fee / (1 - fee)].
+     *
+     *    If R_current > R_target (swap token1 to token0):
+     *      a) R_target = [(amount1 - amountIn)] / [(amount0 + amoutOut) * sqrtPrice² + (amount1 - amountIn)].
+     *      b) amountOut = (1 - fee) * amountIn / sqrtPrice².
+     *         => amountIn = [(R_current - R_target) * (amount0 * sqrtPrice² + amount1)] / (1 - R_target * fee).
+     */
+    function _getSwapParams(
+        uint256 sqrtPrice,
+        uint256 sqrtRatioLower,
+        uint256 sqrtRatioUpper,
+        uint256 balance0,
+        uint256 balance1,
+        uint256 fee
+    ) internal pure returns (bool zeroToOne, uint256 amountIn, uint256 amountOut) {
+        if (sqrtPrice >= sqrtRatioUpper) {
+            // New position is out of range and fully in token 1.
+            // Rebalance to a single-sided liquidity position in token 1.
+            zeroToOne = true;
+            amountIn = balance0;
+            amountOut = _getAmountOut(sqrtPrice, true, balance0, fee);
+        } else if (sqrtPrice <= sqrtRatioLower) {
+            // New position is out of range and fully in token 0.
+            // Rebalance to a single-sided liquidity position in token 0.
+            amountIn = balance1;
+            amountOut = _getAmountOut(sqrtPrice, false, balance1, fee);
+        } else {
+            // Get target ratio in token1 terms.
+            uint256 targetRatio = _getTargetRatio(sqrtPrice, sqrtRatioLower, sqrtRatioUpper);
+
+            // Calculate the total position value in token1 equivalent:
+            uint256 token0ValueInToken1 = PricingLogic._getSpotValue(sqrtPrice, true, balance0);
+            uint256 totalValueInToken1 = balance1 + token0ValueInToken1;
```

Comment:
Can these be also in the unchecked block? At first glance, I did not see any overflow issue

Reply:
balance1 + token0ValueInToken1 could in theory overflow.
while it should never happen under normal conditions, if there is a bug somewhere else and a silent overflow it might give issues.
So prefer to have this one checked.